nbs/15_gsa_gls/20-null_simulations/10_gwas_harmonization/README.md

```diff
@@ -135,17 +151,112 @@ mkdir -p _tmp/postprocessing
+Another check is to count how many parts were processed for each random phenotype.
+It should be 22 chromosomes times 10 batches (220):
+```bash
+bash check_job.sh \
+  -i _tmp/postprocessing/ \
+  -p "INFO - Processing imputed random" \
+  -c 220
+
+# which should output:
+# Finished checking [NUMBER_OF_PHENOTYPES] logs:
+#  All jobs finished successfully
+```
+
+
+## Compute set of common variants
+
+For some reason, the postprocessing scripts previously run do not generate exactly the same set of final variants for imputed GWAS.
+Each final GWAS should have the same set of imputed variants, since they were all imputed using GWAS from the same cohort and exactly the same input variants.
+However, this is not the case, and for this null simulations we need all GWAS to have the same set of variants to efficiently compute gene correlations.
+
+So what we do is to read all final GWAS files, generate a set of common variants across all of them, and then align all final GWAS to that.
+
+
+```python
+import pickle
+from pathlib import Path
+import concurrent
+
+import numpy as np
+import pandas as pd
+
+import conf
+
+N_SAMPLES = 50
+
+POST_IMPUTED_DIR = Path(
+  conf.RESULTS["GLS_NULL_SIMS"],
+  "post_imputed_gwas"
+).resolve()
+assert POST_IMPUTED_DIR.exists(), POST_IMPUTED_DIR
+
+input_files = sorted(list(POST_IMPUTED_DIR.glob("*.txt.gz")))
+len(input_files)
+
+# sample files
+np.random.seed(0)
+input_files = np.random.choice(input_files, size=N_SAMPLES, replace=False)
+len(input_files)
+
+# read all GWAS and find a set of common panel_variant_id_values
+def _get_gwas_variants(f):
+    gwas_data = pd.read_table(f, usecols=["panel_variant_id", "zscore"])
+    assert gwas_data["panel_variant_id"].is_unique
+    assert gwas_data.shape == gwas_data.dropna().shape
+    return f.name, set(gwas_data["panel_variant_id"])
+
+common_variants = set()
+last_n_var_ids = -1
+with concurrent.futures.ProcessPoolExecutor(max_workers=conf.GENERAL["N_JOBS"]) as executor:
+    for gwas_file_name, gwas_variants in executor.map(_get_gwas_variants, input_files, chunksize=10):
+        if len(common_variants) == 0:
+            common_variants = gwas_variants
+        else:
+            common_variants = common_variants.intersection(gwas_variants)
+        
+        n_var_ids = len(common_variants)
+        same_previous = n_var_ids == last_n_var_ids
+        last_n_var_ids = n_var_ids
+        print(f"{gwas_file_name}, # common variants: {n_var_ids} (same? {same_previous})", flush=True)
+
+
+with open(POST_IMPUTED_DIR / "common_variant_ids.pkl", 'wb') as f:
+    pickle.dump(common_variants, f, protocol=pickle.HIGHEST_PROTOCOL)
+```
+
+## Save GWAS files using common variants
+
+```bash
+mkdir -p _tmp/common_var_ids
+cat cluster_jobs/15-common_variant_ids_job.sh | bsub
+```
+
+Checks:
+```bash
+bash check_job.sh \
+  -i _tmp/common_var_ids/ \
+  -f '*.out' \
+  -p "Filtering variants: 8325729"
```

Comment:
Maybe include a note about what this flag means?

Reply:
Yes, good point. I've just added it. It's the "success pattern" that has to be found for the script to finish correctly.